File: src/main/java/com/uml/generator/input/JsonFileReader.java
package com.uml.generator.input;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.uml.generator.model.Trace;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Reads Jaeger traces from JSON files.
 * Supports both single files and directories containing multiple JSON files.
 */
public class JsonFileReader implements TraceReader {

    private static final Logger logger = LoggerFactory.getLogger(JsonFileReader.class);

    private final ObjectMapper objectMapper;
    private final List<File> files;

    /**
     * Constructor for reading a single file.
     * 
     * @param filePath path to the JSON file
     */
    public JsonFileReader(String filePath) {
        this.objectMapper = new ObjectMapper();
        this.files = new ArrayList<>();

        File file = new File(filePath);
        if (file.isDirectory()) {
            this.files.addAll(loadJsonFilesFromDirectory(file));
        } else {
            this.files.add(file);
        }
    }

    /**
     * Constructor for reading multiple files.
     * 
     * @param files list of JSON files
     */
    public JsonFileReader(List<File> files) {
        this.objectMapper = new ObjectMapper();
        this.files = new ArrayList<>(files);
    }

    @Override
    public List<Trace> readTraces() throws Exception {
        List<Trace> traces = new ArrayList<>();

        for (File file : files) {
            logger.info("Reading traces from file: {}", file.getAbsolutePath());

            try {
                List<Trace> fileTraces = readTracesFromFile(file);
                traces.addAll(fileTraces);
                logger.info("Successfully read {} trace(s) from {}", fileTraces.size(), file.getName());
            } catch (Exception e) {
                logger.error("Failed to read traces from file: {}", file.getAbsolutePath(), e);
                throw new Exception("Failed to read traces from file: " + file.getName(), e);
            }
        }

        return traces;
    }

    /**
     * Reads traces from a single JSON file.
     * Handles both single trace and array of traces formats.
     * 
     * @param file the JSON file
     * @return list of traces
     * @throws IOException if reading fails
     */
    private List<Trace> readTracesFromFile(File file) throws IOException {
        List<Trace> traces = new ArrayList<>();

        JsonNode rootNode = objectMapper.readTree(file);

        // Check if the JSON is wrapped in a "data" field (Jaeger API format)
        if (rootNode.has("data")) {
            JsonNode dataNode = rootNode.get("data");

            if (dataNode.isArray()) {
                // Array of traces
                for (JsonNode traceNode : dataNode) {
                    Trace trace = objectMapper.treeToValue(traceNode, Trace.class);
                    traces.add(trace);
                }
            } else {
                // Single trace
                Trace trace = objectMapper.treeToValue(dataNode, Trace.class);
                traces.add(trace);
            }
        } else if (rootNode.isArray()) {
            // Direct array of traces
            for (JsonNode traceNode : rootNode) {
                Trace trace = objectMapper.treeToValue(traceNode, Trace.class);
                traces.add(trace);
            }
        } else {
            // Single trace object
            Trace trace = objectMapper.treeToValue(rootNode, Trace.class);
            traces.add(trace);
        }

        return traces;
    }

    /**
     * Loads all JSON files from a directory (non-recursive).
     * 
     * @param directory the directory
     * @return list of JSON files
     */
    private List<File> loadJsonFilesFromDirectory(File directory) {
        try (Stream<Path> paths = Files.walk(directory.toPath(), 1)) {
            return paths
                    .filter(Files::isRegularFile)
                    .filter(p -> p.toString().toLowerCase().endsWith(".json"))
                    .map(Path::toFile)
                    .collect(Collectors.toList());
        } catch (IOException e) {
            logger.error("Failed to load JSON files from directory: {}", directory.getAbsolutePath(), e);
            return new ArrayList<>();
        }
    }

    public int getFileCount() {
        return files.size();
    }
}

File: python_version/jaeger_uml_generator/input/json_file_reader.py
"""JSON file reader for Jaeger traces."""

import json
import logging
from pathlib import Path
from typing import List
from .trace_reader import TraceReader
from ..models import Trace


logger = logging.getLogger(__name__)


class JsonFileReader(TraceReader):
    """Reads Jaeger traces from JSON files or directories."""
    
    def __init__(self, path: str):
        """
        Initialize the JSON file reader.
        
        Args:
            path: Path to a JSON file or directory containing JSON files
        """
        self.path = Path(path)
    
    def read_traces(self) -> List[Trace]:
        """
        Read traces from JSON file(s).
        
        Returns:
            List of Trace objects
        """
        traces = []
        
        if self.path.is_file():
            traces.extend(self._read_file(self.path))
        elif self.path.is_dir():
            traces.extend(self._read_directory(self.path))
        else:
            logger.error(f"Path does not exist: {self.path}")
            raise FileNotFoundError(f"Path not found: {self.path}")
        
        return traces
    
    def _read_file(self, file_path: Path) -> List[Trace]:
        """Read traces from a single JSON file."""
        logger.info(f"Reading trace file: {file_path}")
        
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            traces = self._parse_json_data(data)
            
            # Set source name from filename
            source_name = file_path.stem
            for trace in traces:
                if not trace.source_name:
                    trace.source_name = source_name
            
            logger.info(f"Loaded {len(traces)} trace(s) from {file_path.name}")
            return traces
            
        except json.JSONDecodeError as e:
            logger.error(f"Invalid JSON in file {file_path}: {e}")
            return []
        except Exception as e:
            logger.error(f"Error reading file {file_path}: {e}")
            return []
    
    def _read_directory(self, dir_path: Path) -> List[Trace]:
        """Read traces from all JSON files in a directory."""
        logger.info(f"Reading trace files from directory: {dir_path}")
        
        all_traces = []
        json_files = list(dir_path.glob('*.json'))
        
        if not json_files:
            logger.warning(f"No JSON files found in directory: {dir_path}")
            return all_traces
        
        for json_file in json_files:
            all_traces.extend(self._read_file(json_file))
        
        return all_traces
    
    def _parse_json_data(self, data: dict) -> List[Trace]:
        """
        Parse JSON data into Trace objects.
        Handles different Jaeger JSON formats.
        """
        traces = []
        
        # Format 1: Jaeger API format with "data" field
        if isinstance(data, dict) and 'data' in data:
            trace_list = data['data']
            if isinstance(trace_list, list):
                for trace_data in trace_list:
                    traces.append(Trace.from_dict(trace_data))
        
        # Format 2: Direct array of traces
        elif isinstance(data, list):
            for trace_data in data:
                traces.append(Trace.from_dict(trace_data))
        
        # Format 3: Single trace object
        elif isinstance(data, dict) and 'traceID' in data:
            traces.append(Trace.from_dict(data))
        
        else:
            logger.warning("Unrecognized JSON format")
        
        return traces
